aws/resource_aws_workspaces_directory.go

```diff
@@ -0,0 +1,348 @@
+	log.Printf("[DEBUG] Workspaces directory %q self-sevice permissions are set", d.Id())
+
+	d.Partial(false)
+
+	return resourceAwsWorkspacesDirectoryRead(d, meta)
+}
+
+func resourceAwsWorkspacesDirectoryRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).workspacesconn
+
+	resp, err := conn.DescribeWorkspaceDirectories(&workspaces.DescribeWorkspaceDirectoriesInput{
+		DirectoryIds: []*string{aws.String(d.Id())},
+	})
+	if err != nil {
+		return err
+	}
+	dir := resp.Directories[0]
+
+	d.Set("directory_id", dir.DirectoryId)
+	d.Set("subnet_ids", dir.SubnetIds)
+	d.Set("self_service_permissions", flattenSelfServicePermissions(dir.SelfservicePermissions))
+
+	tags, err := keyvaluetags.WorkspacesListTags(conn, d.Id())
+	if err != nil {
+		return fmt.Errorf("error listing tags: %s", err)
+	}
+
+	if err := d.Set("tags", tags.IgnoreAws().Map()); err != nil {
+		return fmt.Errorf("error setting tags: %s", err)
+	}
+
+	return nil
+}
+
+func resourceAwsWorkspacesDirectoryUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).workspacesconn
+
+	d.Partial(true)
+
+	if d.HasChange("self_service_permissions") {
+		log.Printf("[DEBUG] Modifying workspaces directory %q self-service permissions...", d.Id())
+		permissions := d.Get("self_service_permissions").([]interface{})
+
+		_, err := conn.ModifySelfservicePermissions(&workspaces.ModifySelfservicePermissionsInput{
+			ResourceId:             aws.String(d.Id()),
+			SelfservicePermissions: expandSelfServicePermissions(permissions),
+		})
+		if err != nil {
+			return fmt.Errorf("error updating self service permissions: %s", err)
+		}
+		log.Printf("[DEBUG] Workspaces directory %q self-sevice permissions are set", d.Id())
+		d.SetPartial("self_service_permission")
+	}
+
+	if d.HasChange("tags") {
+		log.Printf("[DEBUG] Modifying workspaces directory %q tags...", d.Id())
+		o, n := d.GetChange("tags")
+		if err := keyvaluetags.WorkspacesUpdateTags(conn, d.Id(), o, n); err != nil {
+			return fmt.Errorf("error updating tags: %s", err)
+		}
+		log.Printf("[DEBUG] Workspaces directory %q tags are modified", d.Id())
+		d.SetPartial("tags")
+	}
+
+	d.Partial(false)
+
+	return resourceAwsWorkspacesDirectoryRead(d, meta)
+}
+
+func resourceAwsWorkspacesDirectoryDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).workspacesconn
+
+	input := &workspaces.DeregisterWorkspaceDirectoryInput{
+		DirectoryId: aws.String(d.Id()),
+	}
+
+	log.Printf("[DEBUG] Deregestering workspaces directory...\n%#v\n", *input)
+	_, err := conn.DeregisterWorkspaceDirectory(input)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Waiting for workspaces directory %q to become deregistered...", d.Id())
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{
+			workspaces.WorkspaceDirectoryStateRegistering,
+			workspaces.WorkspaceDirectoryStateRegistered,
+		},
+		Target: []string{
+			workspaces.WorkspaceDirectoryStateDeregistering,
```

Comment:
The state `workspaces.WorkspaceDirectoryStateDeregistering` should be in the `Pending` list